### PR TITLE
fix(centos) centos doesn't have su-exec

### PR DESCRIPTION
**Pull request by hutchic**
Tested locally:
```
docker build --no-cache -t kong:centosfix .
export KONG_DOCKER_TAG=kong:centosfix
cd ../compose
docker-compose up -d
docker-compose logs kong

docker-compose exec kong ps aux
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
kong         1  1.0  0.2 266504 18740 ?        Ss   18:15   0:00 nginx: master p
kong        37  0.0  0.4 285084 33700 ?        S    18:15   0:00 nginx: worker p
kong        38  0.0  0.4 285084 33548 ?        S    18:15   0:00 nginx: worker p
root        39  0.0  0.0  51752  3484 ?        Rs+  18:16   0:00 ps aux
```

Manual Test Steps:

- [x] Image builds
- [x] Works in docker-compose
- [x] NGINX master process is PID 1
- [x] NGINX master process is ran by `kong` not `root`
- [x] Works in k8s with transparent proxying

### CODE REVIEW

**Comment by thibaultcha** [comment]
Unfortunately, the container's PID 1 is `su` with this entrypoint:
```
[root@c38f0db99968 /]# ps -ef
UID        PID  PPID  C STIME TTY          TIME CMD
root         1     0  0 21:00 ?        00:00:00 su -c /usr/local/openresty/nginx/sbin/nginx       -p '/usr/local/kong'       -c nginx.conf kong
kong        31     1  1 21:00 ?        00:00:00 nginx: master process /usr/local/openresty/nginx/sbin/nginx -p /usr/local/kong -c nginx.conf
kong        38    31  0 21:00 ?        00:00:00 nginx: worker process
kong        39    31  0 21:00 ?        00:00:00 nginx: worker process
kong        40    31  0 21:00 ?        00:00:00 nginx: worker process
kong        41    31  0 21:00 ?        00:00:00 nginx: worker process
kong        42    31  0 21:00 ?        00:00:00 nginx: worker process
kong        43    31  0 21:00 ?        00:00:00 nginx: worker process
kong        44    31  0 21:00 ?        00:00:00 nginx: worker process
kong        45    31  0 21:00 ?        00:00:00 nginx: worker process
```

**Comment by hutchic** [comment]
@gszr yeah in a future PR I wanted 1.0-centos fixed first :)

**Comment by hutchic** [comment]
@thibaultcha will that matter? I thought so long as the nginx processes themselves were running as the Kong user it was fine?

**Comment by thibaultcha** [comment]
@hutchic Yes, we have always be extra careful for the PID 1 process of our docker image to be the nginx master process. That is by design. If the master process crashes, the container should exit.

File: centos/Dockerfile [hunk 1]
```diff
@@ -3,6 +3,18 @@ LABEL maintainer="Kong Core Team <team-core@konghq.com>"
...
+
+RUN yum install -y -q gcc make \
+&& curl -sL "${SU_EXEC_URL}" | tar -C /tmp -zxf - \ 
+ && make -C "/tmp/su-exec-${SU_EXEC_VERSION}" \
+&& cp "/tmp/su-exec-${SU_EXEC_VERSION}/su-exec" /usr/bin \
+&& rm -fr "/tmp/su-exec-${SU_EXEC_VERSION}" \
+&& yum autoremove -y -q gcc make \
+&& yum clean all -q \
+&& rm -fr /var/cache/yum/* /tmp/yum_save*.yumtx /root/.pki
+
```
Comment (hbagdi) [inline centos/Dockerfile]: I can confirm that this does solve the original problem and also runs nginx master process as PID 1.

Regarding, building su-exec from source, should we build it once and host it on bintray/s3 and use it? Or, we could use [multi-stage build](https://docs.docker.com/develop/develop-images/multistage-build/) to build su-exec and copy it in the production image.

File: centos/Dockerfile [hunk 2]
```diff
@@ -3,6 +3,18 @@ LABEL maintainer="Kong Core Team <team-core@konghq.com>"
...
 
+ARG SU_EXEC_VERSION=0.2
+ARG SU_EXEC_URL="https://github.com/ncopa/su-exec/archive/v${SU_EXEC_VERSION}.tar.gz"
+
+RUN yum install -y -q gcc make \
+&& curl -sL "${SU_EXEC_URL}" | tar -C /tmp -zxf - \ 
+ && make -C "/tmp/su-exec-${SU_EXEC_VERSION}" \
+&& cp "/tmp/su-exec-${SU_EXEC_VERSION}/su-exec" /usr/bin \
+&& rm -fr "/tmp/su-exec-${SU_EXEC_VERSION}" \
+&& yum autoremove -y -q gcc make \
```
Comment (hbagdi) [inline centos/Dockerfile]: I'm not sure why but this doesn't actually remove `make`:
```
sh-4.2# yum list installed | grep make
make.x86_64                               1:3.82-23.el7                @base
```

Running the same command on a centos:7 docker image returns no results.

File: centos/Dockerfile (hunk 1 above)
Comment (hutchic) [reply centos/Dockerfile]: So far as building the rpm / yum asset there's these:
https://github.com/ncopa/su-exec/pull/4
https://bugzilla.redhat.com/show_bug.cgi?id=1393044

I think I'd be more inclined to make our own docker base image

File: centos/Dockerfile (hunk 2 above)
Comment (hutchic) [reply centos/Dockerfile]: lol I figured it out

```
yum install -y https://bintray.com/SNIP/kong-community-edition-0.15.0.el7.noarch.rpm

.....

Dependency Installed:
  make.x86_64 1:3.82-23.el7
  openssl.x86_64 1:1.0.2k-16.el7
```

File: centos/Dockerfile (hunk 1 above)
Comment (hbagdi) [reply centos/Dockerfile]: Not a blocker but only sharing thoughts/practices.

File: centos/Dockerfile [hunk 3]
```diff
@@ -3,6 +3,18 @@ LABEL maintainer="Kong Core Team <team-core@konghq.com>"
 
 ENV KONG_VERSION 0.15.0
 
+ARG SU_EXEC_VERSION=0.2
+ARG SU_EXEC_URL="https://github.com/ncopa/su-exec/archive/v${SU_EXEC_VERSION}.tar.gz"
+
+RUN yum install -y -q gcc make \
+&& curl -sL "${SU_EXEC_URL}" | tar -C /tmp -zxf - \ 
+ && make -C "/tmp/su-exec-${SU_EXEC_VERSION}" \
```
Comment (thibaultcha) [inline centos/Dockerfile]: we have a strange extra space here

**Pull request merged**